### ISSUE TITLE
Registration to template results in blank images when input anat image has mismatched affines

### ISSUE THREAD

**Issue by joshuacwnewton** [spinalcordtoolbox/spinalcordtoolbox#4184]
A user on the forum reported [empty registration results](https://forum.spinalcordmri.org/t/registration-tips/1102/4?u=joshuacwnewton), and @jcohenadad identified [mismatched sform/qform matrices](https://forum.spinalcordmri.org/t/registration-tips/1102/7?u=joshuacwnewton) as a potential cause. I was able to [confirm this](https://forum.spinalcordmri.org/t/registration-tips/1102/8?u=joshuacwnewton), and now I want to identify _why_ the mismatch causes this issue.

So far, I have drawn the following conclusions:

- The issue occurs when checking out the `5.8` tag, so at the very least, this isn't a new issue for v6.0 (last ~6mo of development).
- ~~The issues begin (at the very least) with the straightening step. (I tried using `warp_curve2straight.nii.gz` on the anat image for both cases (matching xforms vs. mismatched xforms), and the output of `sct_apply_transfo` is blank for the mismatched case but correct for the matched case.)~~ This was a red herring. The straightening actually [occurs OK](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4184#issuecomment-1670042547), since the affine mismatch gets fixed internally prior to the straightening step.

Potentially related to: 

- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/2429

**Comment by joshuacwnewton** [comment]
Potentially related to: 

- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/2429

**Comment by joshuacwnewton** [comment]
I'm about to turn in for the day, but I want to quickly summarize where I'm at:

- Thanks to [previous](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/2429) [efforts](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2858), it's currently mandatory that the qform/sform must match during straightening. `sct_straighten_spinalcord` _should_ throw an error when loading any input images with mismatched xforms, thanks to setting `check_sform=True`:
   https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/f5de501d3fe98c53861b54c60b9f5f5b96732772/spinalcordtoolbox/straightening.py#L108-L110
- However, no error is thrown, even when we know that the input images have mismatched s/qforms. Why?
- Looking at the temp folder for `sct_register_to_template`, we see a bunch of preprocessed images:
   <details>

   ![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/ff92d524-18de-4774-a323-64e05b7948b0)
   </details>
- If pick one set of images (e.g. `data.nii`) then check the headers for each, we see that there is a mismatch for `data.nii` and `data_1mm.nii`, but _not_ for `data_1mm_rpi.nii`:
    <details>
    
    ```console
    joshua@tadpole:/tmp/sct_2023-08-07_15-43-15_register-to-template_nxu9sgbd$ sct_image -header -i data.nii 
    qform_name	Scanner Anat
    qform_code	1
    qto_xyz:1	0.000000 0.000000 1.000000 -95.500000 
    qto_xyz:2	-1.000000 0.000000 0.000000 146.947952 
    qto_xyz:3	0.000000 1.000000 0.000000 -63.840141 
    qto_xyz:4	0.000000 0.000000 0.000000 1.000000 
    qform_xorient	Anterior-to-Posterior
    qform_yorient	Inferior-to-Superior
    qform_zorient	Left-to-Right
    sform_name	Aligned Anat
    sform_code	2
    sto_xyz:1	0.000000 0.000000 1.000000 -94.804344 
    sto_xyz:2	-1.000000 0.000000 0.000000 116.308746 
    sto_xyz:3	0.000000 1.000000 0.000000 -174.108688 
    sto_xyz:4	0.000000 0.000000 0.000000 1.000000 
    sform_xorient	Anterior-to-Posterior
    sform_yorient	Inferior-to-Superior
    sform_zorient	Left-to-Right
    joshua@tadpole:/tmp/sct_2023-08-07_15-43-15_register-to-template_nxu9sgbd$ sct_image -header -i data_1mm.nii
    qform_name	Scanner Anat
    qform_code	1
    qto_xyz:1	0.000000 0.000000 1.000000 -95.500000 
    qto_xyz:2	-1.000000 0.000000 0.000000 146.947952 
    qto_xyz:3	0.000000 1.000000 0.000000 -63.840141 
    qto_xyz:4	0.000000 0.000000 0.000000 1.000000 
    qform_xorient	Anterior-to-Posterior
    qform_yorient	Inferior-to-Superior
    qform_zorient	Left-to-Right
    sform_name	Aligned Anat
    sform_code	2
    sto_xyz:1	0.000000 0.000000 1.000000 -94.804344 
    sto_xyz:2	-1.000000 0.000000 0.000000 116.308746 
    sto_xyz:3	0.000000 1.000000 0.000000 -174.108688 
    sto_xyz:4	0.000000 0.000000 0.000000 1.000000 
    sform_xorient	Anterior-to-Posterior
    sform_yorient	Inferior-to-Superior
    sform_zorient	Left-to-Right
    joshua@tadpole:/tmp/sct_2023-08-07_15-43-15_register-to-template_nxu9sgbd$ sct_image -header -i data_1mm_rpi.nii
    qform_name	Scanner Anat
    qform_code	1
    qto_xyz:1	-1.000000 0.000000 0.000000 96.195656 
    qto_xyz:2	0.000000 1.000000 0.000000 -138.691254 
    qto_xyz:3	0.000000 0.000000 1.000000 -174.108688 
    qto_xyz:4	0.000000 0.000000 0.000000 1.000000 
    qform_xorient	Right-to-Left
    qform_yorient	Posterior-to-Anterior
    qform_zorient	Inferior-to-Superior
    sform_name	Aligned Anat
    sform_code	2
    sto_xyz:1	-1.000000 0.000000 0.000000 96.195656 
    sto_xyz:2	0.000000 1.000000 0.000000 -138.691254 
    sto_xyz:3	0.000000 0.000000 1.000000 -174.108688 
    sto_xyz:4	0.000000 0.000000 0.000000 1.000000 
    sform_xorient	Right-to-Left
    sform_yorient	Posterior-to-Anterior
    sform_zorient	Inferior-to-Superior
    ```
    </details>
- And, when we check the sform of the input image, we're checking that final `_1mm_rpi` image. Hence, no error is thrown!
- So, somewhere between the resampling and the reorientation, the sform/qform mismatch is eliminated, even though clearly the initial mismatch has an effect on the straightening!!

I will need to investigate further why the mismatch is getting papered over, and how that changes the straightening results.

**Comment by joshuacwnewton** [comment]
It took me a bit, but here is comprehensive investigation into how the various qform/sform matrices change during registration: 

<details>

(For brevity, I'm going to only reference the x translation value of the matrices.)

1. Pre-registration steps
   - The anat image has mismatched qform (`-95.500`) and sform (`-94.804`).
   - The seg image, however, uses the sform (`-94.804`) for both affines, as noticed by @jcohenadad in [this comment](https://forum.spinalcordmri.org/t/registration-tips/1102/7?u=joshuacwnewton).
   - The label image also uses the sform matrix for both affines (`-94.804`)
2. Preprocessing within `sct_register_to_template`
   - After RPI reorientation, **all 3 temp input images (anat, seg, label) end up with the same qform/sform** (`-96.195`)
   - This is because the `change_orientation` function uses [`get_best_affine()`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/f5de501d3fe98c53861b54c60b9f5f5b96732772/spinalcordtoolbox/image.py#L1110)  to compute the post-reorientation affines. `get_best_affine()` returns the sform (`-94.804`), matmul occurs, and then this new matrix [gets assigned to both qform and sform](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/f5de501d3fe98c53861b54c60b9f5f5b96732772/spinalcordtoolbox/image.py#L1116-L1117). 
   - So, the reoriented RPI anat image (called `ftmp_data`) has its mismatch fixed, and thus the `check_sform` assertion passes.
   - However, the fixed RPI image is not used consistently in  `sct_register_to_template`. In some places, `ftmp_data` (with matching affines) is used, while in other places, `"data.nii"` (with mismatched affines) is used.
   - Whether or not the "fixed" anat image is used is the key!
3. Post-straightening warping and viewing.
   - After straightening, we get sform-based warping fields.
   - If we apply the sform-based `warp_straight2curve.nii.gz` to the mismatched anat image, we get a blank image. 
   - But, if we modify the anat image so that both affine matrices equal the sform, then apply the sform-based warping field, we get a perfectly fine straightened image. 
   - This implies that the straightening actually went fine, and the issues only occur when we try to apply the generated warping fields using the mismatched anat image as a destination image.
   - Importantly, the straightening uses the "fixed" `ftmp_data` image.
4. Post-registration warping and viewing
   - After registration, we get sform-based warping fields.
   - If we apply the sform-based `warp_anat2template.nii.gz` and `warp_template2anat.nii.gz` using the mismatched anat image, we get blank images.
   - But, if we use the `warp_anat2template.nii.gz` file on the "fixed" anat image, it warps just finet to the PAM50 space.
      - This is because the "fixed" anat image is used when generating `warp_anat2template.nii.gz` (see [here](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/f5de501d3fe98c53861b54c60b9f5f5b96732772/spinalcordtoolbox/scripts/sct_register_to_template.py#L723)).
   - However, the `warp_template2anat.nii.gz` is completely wrong. It's _tiny_ (146.9kB) compared to the size it should be when the affines match (5.4MB).
      - This is because the mismatched anat image is used when generating `warp_template2anat.nii.gz` (see [here](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/f5de501d3fe98c53861b54c60b9f5f5b96732772/spinalcordtoolbox/scripts/sct_register_to_template.py#L745)). 

</details>

tl;dr:

- The mismatch in the anat image gets fixed inside of `sct_register_to_template` during reorientation, causing the `check_sform` assertion to pass for the `ftmp_data` image.
- However, this "fixed" anat image is not used consistently. In some places (e.g. straightening, anat->template concatenation) the "fixed" anat image is used, but in other places (e.g. template->anat concatenation) the mismatched anat image is used instead.
- So, the fact that the `check_sform` assertion passes is actually misleading!!
- As a result, we are left with a mix of correct (straightening, anat->template) and incorrect (template->anat) warping fields.
- Separately, there is another issue inside of `sct_apply_transfo`. Even if a "correct" sform-based warping field is applied, the warped image will come out all wrong if there is an affine mismatch for the anat image.

Setting sform == qform fixes all of these issues, of course. But, I would suggest one of the following two options:

   1. _Properly_ utilise `check_sform` to raise an error during registration when the sform/qforms of the anat image don't match.
   2. Commit to using `nibabel`'s affine logic internally (e.g. by more heavily adopting `get_best_affine()`), and ensure that the "best affine" is _consistently_ used in both `sct_register_to_template` and `sct_apply_transfo`, even when the original anat image has mismatching affines.

The first option is small in scope and would be the safest change to implement, but may be an inconvenience for our users. The second option would be smoother for our users, however it may have unintended consequences due to the changes being larger in scope (especially re: `sct_apply_transfo`).

I'm leaning towards 1. for now. While standardizing our affine approach (re: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3283) is appealing, I also recognize that we have a lot of irons in the fire right now (re: suggestions from last SCT meeting, documentation changes for the upcoming SCT course, etc.). So, I'm leaning towards doing a small fix now, and then later prioritizing affine-related refactors for a hypothetical SCT v7.0.
